cpp/oneapi/dal/backend/primitives/rng/dpc_engine_dpc.cpp

```diff
@@ -0,0 +1,186 @@
+
+namespace bk = oneapi::dal::backend;
+
+template <typename Type, typename Size, engine_method EngineType>
+void uniform(sycl::queue& queue,
+             Size count,
+             Type* dst,
+             dpc_engine<EngineType>& engine_,
+             Type a,
+             Type b,
+             const event_vector& deps) {
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) == sycl::usm::alloc::host) {
+        throw domain_error(dal::detail::error_messages::unsupported_data_type());
+    }
+    oneapi::mkl::rng::uniform<Type> distr(a, b);
+    auto event = oneapi::mkl::rng::generate(distr, engine_.get_gpu_engine(), count, dst, { deps });
+    event.wait_and_throw();
+    engine_.skip_ahead_cpu(count);
+}
+
+//Currently only CPU impl
+template <typename Type, typename Size, engine_method EngineType>
+void uniform_without_replacement(sycl::queue& queue,
+                                 Size count,
+                                 Type* dst,
+                                 Type* buffer,
+                                 dpc_engine<EngineType>& engine_,
+                                 Type a,
+                                 Type b,
+                                 const event_vector& deps) {
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) ==
+        sycl::usm::alloc::device) {
+        throw domain_error(dal::detail::error_messages::unsupported_data_type());
+    }
+    void* state = engine_.get_host_engine_state();
+    engine_.skip_ahead_gpu(count);
+    uniform_dispatcher::uniform_without_replacement_by_cpu<Type>(count, dst, buffer, state, a, b);
+}
+
+//Currently only CPU impl
+template <typename Type, typename Size, engine_method EngineType>
+void shuffle(sycl::queue& queue,
+             Size count,
+             Type* dst,
+             dpc_engine<EngineType>& engine_,
+             const event_vector& deps) {
+    Type idx[2];
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) ==
+        sycl::usm::alloc::device) {
+        throw domain_error(dal::detail::error_messages::unsupported_data_type());
+    }
+    void* state = engine_.get_host_engine_state();
+    engine_.skip_ahead_gpu(count);
+
+    for (Size i = 0; i < count; ++i) {
+        uniform_dispatcher::uniform_by_cpu<Type>(2, idx, state, 0, count);
+        std::swap(dst[idx[0]], dst[idx[1]]);
+    }
+}
+
+#define INSTANTIATE_(F, Size, EngineType)                                \
+    template ONEDAL_EXPORT void uniform(sycl::queue& queue,              \
+                                        Size count_,                     \
+                                        F* dst,                          \
+                                        dpc_engine<EngineType>& engine_, \
+                                        F a,                             \
+                                        F b,                             \
+                                        const event_vector& deps);
+
```

Comment:
It would be better to just add `uniform_without_replacement` and `shuffle` instantiations to this macros.
You'll not need `INSTANTIATE_uniform_without_replacement_FLOAT` and other similar macros then.